proof/refine/ARM/Ipc_R.thy

```diff
@@ -2304,20 +2297,454 @@ crunches doIPCTransfer, possibleSwitchTo
+  apply (clarsimp simp: obj_at'_def projectKOs)
+  apply (intro conjI impI; (fastforce simp: obj_at'_def projectKOs valid_tcb_state'_def)?)
+  by (insert reply_ko_at_valid_objs_valid_reply';
+      fastforce simp: valid_reply'_def obj_at'_def projectKOs valid_bound_obj'_def)+
+
+lemma replyPush_valid_queues[wp]:
+  "\<lbrace>valid_queues and valid_objs'\<rbrace>
+   replyPush callerPtr calleePtr replyPtr canDonate
+   \<lbrace>\<lambda>_. valid_queues\<rbrace>"
+  supply if_split [split del]
+  unfolding replyPush_def updateReply_def bind_assoc
+  apply (rule hoare_seq_ext[OF _ stateAssert_inv])
+  apply simp
+  apply (rule hoare_seq_ext[OF _ threadGet_sp'])
+  apply (rule hoare_seq_ext[OF _ threadGet_sp'])
+  apply (rule hoare_seq_ext[OF _ get_reply_sp'])
+  apply (wpsimp wp: schedContextDonate_valid_queues
+                    threadGet_wp hoare_vcg_if_lift2 hoare_vcg_imp_lift')
+  apply (clarsimp simp: ko_at_obj_at'[where P=\<top>])
+  apply (clarsimp simp: valid_reply'_def dest!: reply_ko_at_valid_objs_valid_reply')
+  done
+
+crunches reply_unlink_tcb
+  for sc_replies_sc_at[wp]: "\<lambda>s. Q (sc_replies_sc_at P scp s)"
+  (wp: crunch_wps)
+
+(* FIXME RT: move to AInvs *)
+lemma BlockedOnReply_sym_ref1:
+  "\<lbrakk>st_tcb_at ((=) (Structures_A.thread_state.BlockedOnReply rp)) thread s; sym_refs (state_refs_of s)\<rbrakk>
+  \<Longrightarrow> reply_tcb_reply_at ((=) (Some thread)) rp s"
+  apply (clarsimp simp: pred_tcb_at_def reply_tcb_reply_at_def)
+  apply (drule (1) sym_refs_obj_atD[where p=thread])
+  apply (clarsimp simp: state_refs_of_def obj_at_def tcb_st_refs_of_def
+                 split: Structures_A.thread_state.splits)
+  apply (rename_tac ko; case_tac ko; clarsimp simp: get_refs_def2)
+  done
+
+(* FIXME RT: move to AInvs *)
+lemma BlockedOnReply_sym_ref2:
+  "\<lbrakk>sym_refs (state_refs_of s);
+    valid_replies s;
+    reply_tcb_reply_at ((=) (Some tcbPtr)) rPtr s;
+    (rPtr, scp') \<in> replies_with_sc s\<rbrakk> \<Longrightarrow>
+    st_tcb_at (\<lambda>st. st = Structures_A.thread_state.BlockedOnReply rPtr) tcbPtr s"
+  apply (frule (1) valid_repliesD1_simp, clarsimp simp: replies_blocked_def)
+  apply (subgoal_tac "t = tcbPtr")
+   apply (clarsimp simp: pred_tcb_at_def obj_at_def)
+  apply (subgoal_tac "reply_tcb_reply_at ((=) (Some t)) rPtr s")
+   apply (clarsimp simp: reply_tcb_reply_at_def obj_at_def)
+   apply (case_tac "reply_tcb r"; clarsimp)
+  apply (rule BlockedOnReply_sym_ref1; clarsimp simp: pred_tcb_at_def obj_at_def)
+  done
+
+crunches doIPCTransfer
+  for reply_projs[wp]: "\<lambda>s. P (replyNexts_of s) (replyPrevs_of s) (replyTCBs_of s) (replySCs_of s)"
+  (wp: crunch_wps simp: crunch_simps)
+
+lemma BlockedOnReceive_reply_tcb_reply_at:
+  "\<lbrakk>\<exists>epptr pl. st_tcb_at ((=) (Structures_A.BlockedOnReceive epptr (Some rptr) pl)) tptr s;
+    sym_refs (state_refs_of s); valid_objs s\<rbrakk>
+   \<Longrightarrow> reply_tcb_reply_at (\<lambda>tp_opt. tp_opt = Some tptr) rptr s"
```

Comment:
> This looks exactly the same as `DetSchedInvs_AI.BlockedOnReceive_reply_tcb_reply_at` (including the proof)...

So, if it is actually the same, we don’t need this lemma here, I suppose? And probably update the one in AInvs so that it doesn’t use `valid_objs`?

Reply:
I would not be surpised if this lemma exists, but I can't find it. Can you give me a line number in file in a branch?

Reply:
Hi Mitch. Sorry for the confusion, but this lemma has been actually removed in #255 . I was on an old branch. So, this was replaced by `st_tcb_recv_reply_state_refs`. If you look at #255 , it should be pretty straightforward to use `st_tcb_recv_reply_state_refs` instead of this one.